MOLT Fetch GA documentation (#18541)
cockroachdb/docs: src/current/v24.1/molt-fetch.md

docs_area: migrate
---

> {{site.data.alerts.callout_info}}

Reviewer: Do you need to remove it from the Feature Phases page?

> MOLT Fetch uses [`IMPORT INTO`]({% link {{ page.version.version }}/import-into.md %}) or [`COPY FROM`]({% link {{ page.version.version }}/copy-from.md %}) to move the source data to cloud storage (Google Cloud Storage or Amazon S3), a local file server, or local memory. Once the data is exported, MOLT Fetch loads the data onto a target CockroachDB database. For details, see [Usage](#usage).

Reviewer: No Azure?

Reviewer 2: Currently we don't support Azure, but theoretically shouldn't be too hard to implement and test. Just need to get around to doing it.

Complete the following items before using MOLT Fetch:

- Follow the recommendations in [Best practices](#best-practices).

- Ensure that the source and target schemas are identical, unless you enable automatic schema creation with the [`'drop-on-target-and-recreate'`](#target-table-handling) option.

- Ensure that the SQL user running MOLT Fetch has [`SELECT` privileges]({% link {{ page.version.version }}/grant.md %}#supported-privileges) on the source and target CockroachDB databases, along with the required privileges to run [`IMPORT INTO`]({% link {{ page.version.version }}/import-into.md %}#required-privileges) or [`COPY FROM`]({% link {{ page.version.version }}/copy-from.md %}#required-privileges) (depending on the [fetch mode](#fetch-mode)) on CockroachDB, as described on their respective pages.

> - To enable continuous replication using [`--ongoing-replication`](#replication) or the [CDC cursor](#cdc-cursor):

Reviewer 3: This sentence reads a bit confusing as to what to actually do.

			~~~

		- Ensure the Google Cloud Storage bucket is created and accessible to CockroachDB.

## Best practices

> - To prevent connections from prematurely terminating during data export, set the following to high values on the source database:

Reviewer: ```suggestion
- To prevent connections from terminating prematurely during data export, set the following to high values on the source database:
```
Nit to avoid splitting the infinitive, but I think it reads more clearly too

> 	- **Maximum allowed number of connections.** MOLT Fetch can export data across multiple connections. The number of connections it will create is the number of shards ([`--export-concurrency`](#global-flags)) multiplied by the number of tables ([`--table-concurrency`](#global-flags)) being exported concurrently.

Reviewer: It looks weird to me that these have periods. I think maybe they should have colons after the bold instead.

	- **Maximum lifetime of a connection.** This is particularly important for MySQL sources, which can only use a single connection to move data. See the following note.

> - If a MySQL database is set as a [source](#source-and-target-databases), the [`--table-concurrency`](#global-flags) and [`--export-concurrency`](#global-flags) flags **cannot** be set above `1`. If these values are changed, MOLT Fetch returns an error. This is required in order to guarantee consistency when moving data from MySQL, due to MySQL limitations. MySQL data is migrated to CockroachDB one table and shard at a time, using [`WITH CONSISTENT SNAPSHOT`](https://dev.mysql.com/doc/refman/8.0/en/commit.html) transactions.

Reviewer: ```suggestion
- If a MySQL database is set as a [source](#source-and-target-databases), the [`--table-concurrency`](#global-flags) and [`--export-concurrency`](#global-flags) flags **cannot** be set above `1`. If these values are changed, MOLT Fetch returns an error. This helps to guarantee consistency when moving data from MySQL, due to MySQL limitations. MySQL data is migrated to CockroachDB one table and shard at a time, using [`WITH CONSISTENT SNAPSHOT`](https://dev.mysql.com/doc/refman/8.0/en/commit.html) transactions.
```

## Commands

| Command |                                               Usage                                               |
|---------|---------------------------------------------------------------------------------------------------|
| `fetch` | Start the fetch process. This loads data from a source database to a target CockroachDB database. |

### Subcommands

> The following subcommands are run after the `fetch` command.

Reviewer: Do you mean that the following subcommands are supported by the `fetch` command? It sort of sounds like they happen automatically when you run `fetch` the way it is now.

|   Command    |                                Usage                                 |
|--------------|----------------------------------------------------------------------|

> | `token list` | List active [continuation tokens](#list-active-continuation-tokens). |

Reviewer 2: Slight correction here: `tokens list` instead of `token list`

| `--continuation-token`                        | Restart fetch at a specific table, using the specified continuation token, if the process encounters an error. `--fetch-id` must be specified. For details, see [Fetch continuation](#fetch-continuation).                                                                                                                                                                                                                |
| `--direct-copy`                               | Enables [direct copy mode](#fetch-mode), which copies data directly from source to target without using an intermediate store.                                                                                                                                                                                                                                                                                            |

> | `--export-concurrency`                        | Number of shards to export at a time, each on a dedicated thread. **Note:** The number of concurrent threads is the product of `--export-concurrency` and `--table-concurrency`. See [Best practices](#best-practices).<br><br>This value **cannot** be set higher than `1` when moving data from MySQL. See [Best practices](#best-practices).<br><br>**Default:** `4` with a PostgreSQL source; `1` with a MySQL source |

Reviewer: ```suggestion
| `--export-concurrency`                        | Number of shards to export at a time, each on a dedicated thread. **Note:** The number of concurrent threads is the product of `--export-concurrency` and `--table-concurrency`. See [Best practices](#best-practices).<br><br>This value **cannot** be set higher than `1` when moving data from MySQL. Refer to [Best practices](#best-practices).<br><br>**Default:** `4` with a PostgreSQL source; `1` with a MySQL source |
```

| `--ongoing-replication`                       | Enable continuous [replication](#replication) to begin after the fetch process succeeds (i.e., initial source data is loaded into CockroachDB).                                                                                                                                                                                                                                                                           |
| `--pglogical-replication-slot-drop-if-exists` | Drop the replication slot, if specified with `--pglogical-replication-slot-name`. Otherwise, the default replication slot is not dropped.                                                                                                                                                                                                                                                                                 |

> | `--pglogical-replication-slot-name`           | The name of a replication slot to create before taking a snapshot of data (e.g., `'fetch'`). **Required** in order to perform continuous [replication](#replication) from a source PostgreSQL database.                                                                                                                                                                                                                   |

Reviewer: Do these `pg` ones get ignored if importing from MySQL?

Author: @ZhouXing19 what's the behavior here (if you pass the pglogical flags but have MySQL source)?

Reviewer 4: Yes I think so, but might want to confirm with @ryanluu12345 who has more context here.

Reviewer 3: Yes. It should be ignored since we won't enter the code path that uses these.

| `--pglogical-replication-slot-plugin`         | The output plugin used for logical replication under `--pglogical-replication-slot-name`.<br><br>**Default:** `pgoutput`                                                                                                                                                                                                                                                                                                  |
| `--pprof-listen-addr`                         | Address of the pprof endpoint.<br><br>**Default:** `'127.0.0.1:3031'`                                                                                                                                                                                                                                                                                                                                                     |

> | `--replicator-flags`                          | If continuous [replication](#replication) is enabled with `--ongoing-replication`, specify Replicator arguments to override.                                                                                                                                                                                                                                                                                              |

Reviewer 2: Should we link to replicator documentation wiki for this? https://github.com/cockroachdb/replicator/wiki

This will allow people to know which replicator flags they can pass in ^

| `--use-console-writer`                        | Use the console writer, which has cleaner log output but introduces more latency.<br><br>**Default:** `false` (log as structured JSON)                                                                                                                                                                                                                                                                                    |
| `--use-copy`                                  | Use [`COPY FROM` mode](#fetch-mode) to move data. This makes tables queryable during data load, but is slower than `IMPORT INTO` mode. For details, see [Fetch mode](#fetch-mode).                                                                                                                                                                                                                                        |

> ### `token` flags

Reviewer 2: Should be `token list` flags, since the command is `./molt fetch token list`

--table-handling 'drop-on-target-and-recreate'
~~~

When using the `'drop-on-target-and-recreate'` option, MOLT Fetch creates a new CockroachDB table to load the source data if one does not already exist. To guide schema creation, you can [explicitly map source types to CockroachDB types](#type-mapping).

#### Type mapping

If [`'drop-on-target-and-recreate'`](#target-table-handling) is set, MOLT Fetch automatically creates a CockroachDB schema that is compatible with the source data. The column types are determined as follows:

> - All PostgreSQL types are mapped to identical CockroachDB [types]({% link {{ page.version.version }}/data-types.md %}).

Reviewer 4: ```suggestion
- A PostgreSQL type is mapped to a CockroachDB [type]({% link {{ page.version.version }}/data-types.md %}) with the same type oid.
```

Reviewer 4: Yeah i feel torn here as well. In contrast to `same` or `identical`, maybe `corresponding` would be better?

The tricky thing with PG is that [CRDB](https://www.cockroachlabs.com/docs/stable/data-types) only support a subset of [PG types](https://www.postgresql.org/docs/current/datatype.html). For example, CRDB doesn't have a `money` type from PG.  But for those overlapped by PG and CRDB, most of them are 1-1. 

MySQL is a bit tricky as it can be many-to-one, which makes `same` or `identical` less accurate.
Is it ok if we say that a PG type is mapped to the corresponding CockroachDB type based on the oid, if applicable. For MySQL, since we support a much smaller set of mysql types, I think it's ok for us to exhaust all mapping rules. can we still keep those entries even they are the same for mysql and cockroachdb?

Author: I see! OK, I made the table complete again.

> - Some MySQL types are mapped to different CockroachDB types:

Reviewer 4: ```suggestion
- Below is the default mapping from a MySQL column type to a CockroachDB Type
```

	| `BINARY`, `VARBINARY`                        | [`BYTES`]({% link {{ page.version.version }}/bytes.md %})                                                      |
	| `DATETIME`                                   | [`TIMESTAMP`]({% link {{ page.version.version }}/timestamp.md %})                                              |
	| `TIMESTAMP`                                  | [`TIMESTAMPTZ`]({% link {{ page.version.version }}/timestamp.md %})                                            |
	| `BIT`                                        | [`VARBIT`]({% link {{ page.version.version }}/bit.md %})                                                       |
	| `TINYBLOB`, `MEDIUMBLOB`, `LONGBLOB`         | [`BYTES`]({% link {{ page.version.version }}/bytes.md %})                                                      |
	| `BOOLEAN`                                    | [`BOOL`]({% link {{ page.version.version }}/bool.md %})                                                        |
	| `ENUM`                                       | [`ANY_ENUM`]({% link {{ page.version.version }}/enum.md %})                                                    |

> - Source types can be explicitly mapped to target CockroachDB types using a JSON file, thus overriding the preceding behavior. The allowable type mappings are valid CockroachDB aliases, casts, and the following, which are specific to MOLT Fetch and [Verify]({% link {{ page.version.version }}/molt-verify.md %}):

Reviewer 4: ```suggestion
- If would like to have source types be mapped to CockroachDB types different from the default mapping above for auto schema creation (link to --table-handling=drop-and-recreate-schema), using a JSON file via the `--type-map-file` to override the default mapping. The allowable customized type mappings are valid CockroachDB aliases, casts, and the following, which are specific to MOLT Fetch and [Verify]({% link {{ page.version.version }}/molt-verify.md %}):
```

	- [`VARCHAR`]({% link {{ page.version.version }}/string.md %}) <> [`UUID`]({% link {{ page.version.version }}/uuid.md %})
	- [`BOOL`]({% link {{ page.version.version }}/bool.md %}) <> [`INT2`]({% link {{ page.version.version }}/int.md %})
	- [`VARBIT`]({% link {{ page.version.version }}/bit.md %}) <> [`TEXT`]({% link {{ page.version.version }}/string.md %})
	- [`JSONB`]({% link {{ page.version.version }}/jsonb.md %}) <> [`TEXT`]({% link {{ page.version.version }}/string.md %})
	- [`INET`]({% link {{ page.version.version }}/inet.md %}) <> [`TEXT`]({% link {{ page.version.version }}/string.md %})

`--type-map-file` specifies the path to the JSON file that contains explicit type mappings. For example:

{% include_cached copy-clipboard.html %}
~~~
--type-map-file 'type-mappings.json'
~~~

The JSON should be formatted as follows:

~~~ json
[
  {
    "table": "public.t1",
    "column-type-map": [
      {
        "column": "*",
        "type-kv": {
          "source-type": "int",
          "crdb-type": "INT2"
        }
      },
      {
        "column": "name",
        "type-kv": {
          "source-type": "varbit",
          "crdb-type": "string"
        }
      }
    ]
  }
]
~~~

> 

Reviewer 4: `table` must be of `{schema}.{table}` format

- `type-kv` specifies the `source-type` that maps to the target `crdb-type`.
- `column` specifies the column name to convert. If `*` is specified, then all columns with the matching `source-type` are converted.

### Fetch continuation

If `molt fetch` fails while loading data into CockroachDB from intermediate files, it exits with an error message, fetch ID, and [continuation token](#list-active-continuation-tokens) for each table that failed to load on the target database. You can use this information to continue the process from the *continuation point* where it was interrupted. For an example, see [Continue fetch after encountering an error](#continue-fetch-after-encountering-an-error).

Continuation is only possible under the following conditions:

- All data has been exported from the source database into intermediate files on [cloud](#cloud-storage) or [local storage](#local-file-server).
- The *initial load* of source data to the target CockroachDB database is incomplete. This means that ongoing [replication](#replication) of source data has not begun.

{{site.data.alerts.callout_info}}

> Only one fetch ID and set of continuation tokens are active at any time. See [List active continuation tokens](#list-active-continuation-tokens).

Reviewer 2: Nit: `set of corresponding continuation tokens`. I don't know if it's worth mentioning, but each continuation token maps to a single table

--table-handling 'drop-on-target-and-recreate'
~~~

When using the `'drop-on-target-and-recreate'` option, MOLT Fetch creates a new CockroachDB table to load the source data if one does not already exist. To guide the automatic schema creation, you can [explicitly map source types to CockroachDB types](#type-mapping).

#### Type mapping

> If [`'drop-on-target-and-recreate'`](#target-table-handling) is set, MOLT Fetch automatically creates a CockroachDB schema that is compatible with the source data. The column types are determined as follows:

Reviewer: Sometimes we say "MOLT Fetch" like here, other times we say `molt fetch`. It's a nit, but it might be good to be consistent.

Author: I made a few adjustments to hopefully clarify that `molt fetch` pertains directly to the running process, and MOLT Fetch is describing the functionality more generally.

- PostgreSQL types are mapped to existing CockroachDB [types]({% link {{ page.version.version }}/data-types.md %}) that have the same [`OID`]({% link {{ page.version.version }}/oid.md %}).
- The following MySQL types are mapped to corresponding CockroachDB types:

> 	|                      MySQL type                     |                                                CockroachDB type                                                |

Reviewer: I think if you including leading and trailing `|` characters, you will see a weird artifact on the left-hand side right after the table. This doesn't happen if you remove the leading and trailing pipes, and as a bonus you don't have to maintain the width / alignment of the last column manually. In the style guide, we mention not to use leading and trailing pipes.

Author: Noted... I'm using a plugin that auto-formats the tables emacs-style, which is why these are here & I don't have to worry about manual alignment.

	| `TIMESTAMP`                                         | [`TIMESTAMPTZ`]({% link {{ page.version.version }}/timestamp.md %})                                            |
	| `TIME`                                              | [`TIME`]({% link {{ page.version.version }}/time.md %})                                                        |
	| `BIT`                                               | [`VARBIT`]({% link {{ page.version.version }}/bit.md %})                                                       |
	| `DATE`                                              | [`DATE`]({% link {{ page.version.version }}/date.md %})                                                        |
	| `TINYBLOB`, `MEDIUMBLOB`, `LONGBLOB`                | [`BYTES`]({% link {{ page.version.version }}/bytes.md %})                                                      |
	| `BOOL`, `BOOLEAN`                                   | [`BOOL`]({% link {{ page.version.version }}/bool.md %})                                                        |
	| `ENUM`                                              | [`ANY_ENUM`]({% link {{ page.version.version }}/enum.md %})                                                    |

> - Source types can be explicitly mapped to target CockroachDB types, thus overriding the preceding default mappings for automatic schema creation. These are specified using a JSON file and `--type-map-file`. The allowable custom mappings are valid CockroachDB aliases, casts, and the following mappings specific to MOLT Fetch and [Verify]({% link {{ page.version.version }}/molt-verify.md %}):

Reviewer: ```suggestion
- To override the default mappings for automatic schema creation, you can map source to target CockroachDB types explicitly. These are specified using a JSON file and `--type-map-file`. The allowable custom mappings are valid CockroachDB aliases, casts, and the following mappings specific to MOLT Fetch and [Verify]({% link {{ page.version.version }}/molt-verify.md %}):
```

~~~

### Replication

`--ongoing-replication` enables logical replication from the source database to the target CockroachDB database. 

{% include_cached copy-clipboard.html %}
~~~
--ongoing-replication
~~~

When the `--ongoing-replication` flag is set, changes on the source database are continuously replicated on CockroachDB. This begins only after the fetch process succeeds—i.e., the initial source data is loaded into CockroachDB—as indicated by a `fetch complete` message in the output.

Before using this feature, complete the following:

- Install the Replicator binary. Before running `molt fetch` with continuous replication, [download the binary that matches your system](https://github.com/cockroachdb/replicator/wiki/Installing#automated-builds). The Replicator binary **must** be located in the same directory as your [`molt` binary](#installation).
- Configure the source PostgreSQL or MySQL database for continuous replication, as described in [Setup](#setup).

If the source is a PostgreSQL database, you must also specify a replication slot name:

{% include_cached copy-clipboard.html %}
~~~
--ongoing-replication
--pglogical-replication-slot-name 'replication_slot'
~~~

> If you need to customize the Replicator behavior, use `--replicator-flags` to specify one or more Replicator flags ([PostgreSQL](https://github.com/cockroachdb/replicator/wiki/PGLogical#postgresql-logical-replication) or [MySQL](https://github.com/cockroachdb/replicator/wiki/MYLogical#mysqlmariadb-replication)) to override. This will only be necessary for advanced use cases.

Reviewer: ```suggestion
To customize the Replicator behavior (an advanced use case), use `--replicator-flags` to specify one or more Replicator flags ([PostgreSQL](https://github.com/cockroachdb/replicator/wiki/PGLogical#postgresql-logical-replication) or [MySQL](https://github.com/cockroachdb/replicator/wiki/MYLogical#mysqlmariadb-replication)) to override.
```
This lets most people know to stop reading here

Before using this feature, complete the following:

- Install the Replicator binary. Before running `molt fetch` with continuous replication, [download the binary that matches your system](https://github.com/cockroachdb/replicator/wiki/Installing#automated-builds). The Replicator binary **must** be located in the same directory as your [`molt` binary](#installation).
- Configure the source PostgreSQL or MySQL database for continuous replication, as described in [Setup](#setup).

If the source is a PostgreSQL database, you must also specify a replication slot name:

{% include_cached copy-clipboard.html %}
~~~
--ongoing-replication
--pglogical-replication-slot-name 'replication_slot'
~~~

If you need to customize the Replicator behavior, use `--replicator-flags` to specify one or more Replicator flags ([PostgreSQL](https://github.com/cockroachdb/replicator/wiki/PGLogical#postgresql-logical-replication) or [MySQL](https://github.com/cockroachdb/replicator/wiki/MYLogical#mysqlmariadb-replication)) to override. This will only be necessary for advanced use cases.

{% include_cached copy-clipboard.html %}
~~~
--ongoing-replication

> --replicator-flags '--applyTimeout 1h'

Reviewer 3: I think we might want to show the example of more than 1 flag.
cc @ryanluu12345 to provide example.

Reviewer 2: `--replicator-flags "--applyTimeout '1h' --parallelism 64`

## Examples

> The following examples demonstrate how to issue `molt fetch` commands to load data into CockroachDB. 

Reviewer: ```suggestion
The following examples demonstrate how to issue `molt fetch` commands to import data into CockroachDB. 
```
I think previously we called it importing. I wonder if loading is somehow different. I notice that the subsections below also use "Load".

Author: "Load" is generally used in migration scenarios. My concern about "import" is that it could be conflated with the `IMPORT INTO` command, which is only one of the methods we can apply here.

--table-filter 'employees' \
--bucket-path 'gs://migration/data/cockroach' \
--use-copy \

> --cleanup \

Reviewer 2: No ending `\` if this is the last line of the command.

- `--bucket-path` specifies a directory on an [Google Cloud Storage bucket](#data-path) where intermediate files will be written.
- `--use-copy` specifies that `COPY FROM` is used to load the tables, keeping the source tables online and queryable but loading the data more slowly than `IMPORT INTO`.
- `--cleanup` specifies that the intermediate files should be removed after the source data is loaded.

> - `--ongoing-replication` starts continuous [replication](#replication) of data from the source database to CockroachDB after the fetch process succeeds.

Reviewer: Does it make sense to capture this into a variable or include for reuse? It's repeated verbatim.

To retry a specific table, reissue the initial `molt fetch` command and include the fetch ID and a continuation token:

{{site.data.alerts.callout_success}}

> To list all active continuation tokens, run a `molt fetch token list` command. See [List active continuation tokens](#list-active-continuation-tokens).

Reviewer: ```suggestion
You can use `molt fetch token list` to list all active continuation tokens. Refer to [List active continuation tokens](#list-active-continuation-tokens).
```

--continuation-token '5e7c7173-101c-4539-9b8d-28fad37d0240'
~~~

> To retry all tables that failed, exclude `--continuation-token` from the command. When prompted, type `y` to clear all active continuation tokens. To avoid the prompt (e.g., when running `molt fetch` in a job), include the `--non-interactive` flag:

Reviewer: ```suggestion
To retry all tables with failures, exclude `--continuation-token` from the command. When prompted, type `y` to clear all active continuation tokens. To avoid the prompt (e.g., when running `molt fetch` in a job), include the `--non-interactive` flag:
```
